src/datarobotx/idp/custom_application_source.py

```diff
@@ -0,0 +1,320 @@
+#
+# Copyright 2024 DataRobot, Inc. and its affiliates.
+#
+# All rights reserved.
+#
+# DataRobot, Inc.
+#
+# This is proprietary source code of DataRobot, Inc. and its
+# affiliates.
+#
+# Released under the terms of DataRobot Tool and Utility Agreement.
+# https://www.datarobot.com/wp-content/uploads/2021/07/DataRobot-Tool-and-Utility-Agreement.pdf
+
+# mypy: disable-error-code="attr-defined"
+# pyright: reportPrivateImportUsage=false
+
+import contextlib
+import json
+import os
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from requests_toolbelt import MultipartEncoder
+
+import datarobot as dr
+from datarobot.models.runtime_parameters import RuntimeParameterValue
+from datarobot.utils import camelize
+
+from datarobotx.idp.common.hashing import get_hash
+
+
+def get_or_create_custom_application_source(endpoint: str, token: str, name: str) -> str:
+    """Get or create a custom application source with the given name.
+
+    Args:
+        endpoint: The DataRobot endpoint.
+        token: The DataRobot API token.
+        name: The name of the custom application source.
+
+    Returns
+    -------
+            The ID of the custom application source.
+    """
+    if endpoint is not None and token is not None:
+        client = dr.Client(endpoint=endpoint, token=token)  # type: ignore
+    else:
+        client = dr.client.get_client()
+
+    custom_application_source = client.get("customApplicationSources/").json()["data"]
+
+    for custom_application_source in custom_application_source:
+        if custom_application_source["name"] == name:
+            return str(custom_application_source["id"])
+
+    create_custom_application_source_response = client.post(
+        "customApplicationSources/",
+    )
+    custom_application_source_id = create_custom_application_source_response.json()["id"]
+
+    client.patch(
+        f"customApplicationSources/{custom_application_source_id}",
+        json={"name": name},
+    )
+    return str(custom_application_source_id)
+
+
+# Helper functions
+def _find_existing_custom_application_source_version(
+    endpoint: str, token: str, custom_application_source_id: str, version_token: str
+) -> str:
+    client = dr.Client(endpoint=endpoint, token=token)  # type: ignore
+    versions = client.get(
+        f"customApplicationSources/{custom_application_source_id}/versions/"
+    ).json()["data"]
+    for version in versions:
+        if version.get("label") and version_token in version["label"]:
```

Comment:
No need to change but if you'd like this line to be shorter you could do:
`if version_token in version.get("label", "")`